File: internal/workflow/validation.go
package workflow

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strings"

	"github.com/vamosdalian/kinetic/internal/model/entity"
)

var taskRefPattern = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_]*$`)

func ValidateDefinition(tasks []entity.TaskEntity, edges []entity.EdgeEntity) error {
	taskMap := make(map[string]entity.TaskEntity, len(tasks))
	taskRefMap := make(map[string]string, len(tasks))
	inbound := make(map[string]int, len(tasks))
	outbound := make(map[string][]entity.EdgeEntity, len(tasks))
	indegree := make(map[string]int, len(tasks))

	for _, task := range tasks {
		if task.ID == "" {
			return fmt.Errorf("task id is required")
		}
		ref := task.RefOrDefault()
		if !taskRefPattern.MatchString(ref) {
			return fmt.Errorf("task %s has invalid ref %s", task.NameOrID(), ref)
		}
		if existingTaskID, ok := taskRefMap[ref]; ok && existingTaskID != task.ID {
			return fmt.Errorf("task ref %s must be unique", ref)
		}
		taskRefMap[ref] = task.ID
		taskMap[task.ID] = task
		inbound[task.ID] = 0
		indegree[task.ID] = 0
		outbound[task.ID] = nil

		if err := validateTaskConfig(task); err != nil {
			return fmt.Errorf("task %s: %w", task.NameOrID(), err)
		}
	}

	for _, edge := range edges {
		if _, ok := taskMap[edge.Source]; !ok {
			return fmt.Errorf("edge source %s not found", edge.Source)
		}
		if _, ok := taskMap[edge.Target]; !ok {
			return fmt.Errorf("edge target %s not found", edge.Target)
		}
		inbound[edge.Target]++
		indegree[edge.Target]++
		outbound[edge.Source] = append(outbound[edge.Source], edge)
	}

	for _, task := range tasks {
		if task.Type == "condition" {
			if inbound[task.ID] != 1 {
				return fmt.Errorf("condition task %s must have exactly one incoming edge", task.NameOrID())
			}
			if len(outbound[task.ID]) != 2 {
				return fmt.Errorf("condition task %s must have exactly two outgoing edges", task.NameOrID())
			}

			handles := map[string]bool{}
			for _, edge := range outbound[task.ID] {
				if edge.SourceHandle == "" {
					return fmt.Errorf("condition task %s requires source handles true and false", task.NameOrID())
				}
				handles[edge.SourceHandle] = true
			}
			if !handles["true"] || !handles["false"] {
				return fmt.Errorf("condition task %s requires source handles true and false", task.NameOrID())
			}
			continue
		}

		if task.Type != "for" {
			continue
		}

		bodyCount := 0
		doneCount := 0
		handles := map[string]bool{}
		for _, edge := range outbound[task.ID] {
			if edge.SourceHandle == "" {
				return fmt.Errorf("for task %s requires source handle body", task.NameOrID())
			}
			handles[edge.SourceHandle] = true
			switch edge.SourceHandle {
			case "body":
				bodyCount++
			case "done":
				doneCount++
			default:
				return fmt.Errorf("for task %s only supports source handles body and done", task.NameOrID())
			}
		}
		if !handles["body"] || bodyCount != 1 {
			return fmt.Errorf("for task %s requires exactly one body source handle", task.NameOrID())
		}
		if doneCount > 1 {
			return fmt.Errorf("for task %s supports at most one done source handle", task.NameOrID())
		}
	}

	queue := make([]string, 0, len(tasks))
	for taskID, count := range indegree {
		if count == 0 {
			queue = append(queue, taskID)
		}
	}

	visited := 0
	for len(queue) > 0 {
		taskID := queue[0]
		queue = queue[1:]
		visited++
		for _, edge := range outbound[taskID] {
			indegree[edge.Target]--
			if indegree[edge.Target] == 0 {
				queue = append(queue, edge.Target)
			}
		}
	}

	if visited != len(tasks) {
		return fmt.Errorf("workflow graph contains a cycle")
	}

	return nil
}

func validateTaskConfig(task entity.TaskEntity) error {
	switch task.Type {
	case "shell":
		var cfg ShellConfig
		if err := decodeConfig(task.Config, &cfg); err != nil {
			return fmt.Errorf("invalid shell config: %w", err)
		}
		if strings.TrimSpace(cfg.Script) == "" {
			return fmt.Errorf("shell script is required")
		}
		if err := ValidateTemplateString(cfg.Script); err != nil {
			return err
		}
		if err := ValidateTemplateEnvValues(cfg.TaskPolicy.Env); err != nil {
			return err
		}
	case "http":
		var cfg HTTPConfig
		if err := decodeConfig(task.Config, &cfg); err != nil {
			return fmt.Errorf("invalid http config: %w", err)
		}
		if strings.TrimSpace(cfg.URL) == "" {
			return fmt.Errorf("http url is required")
		}
		if err := ValidateTemplateString(cfg.URL); err != nil {
			return err
		}
		if err := ValidateTemplateString(cfg.Method); err != nil {
			return err
		}
		if err := ValidateTemplateString(cfg.Body); err != nil {
			return err
		}
		for key, value := range cfg.Headers {
			if err := ValidateTemplateString(value); err != nil {
				return fmt.Errorf("invalid template in header %s: %w", key, err)
			}
		}
		if err := ValidateTemplateEnvValues(cfg.TaskPolicy.Env); err != nil {
			return err
		}
	case "condition":
		var cfg ConditionConfig
		if err := decodeConfig(task.Config, &cfg); err != nil {
			return fmt.Errorf("invalid condition config: %w", err)
		}
		if strings.TrimSpace(cfg.Expression) == "" {
			return fmt.Errorf("condition expression is required")
		}
		if err := ValidateTemplateString(cfg.Expression); err != nil {
			return err
		}
		if !ContainsTemplate(cfg.Expression) {
			if _, err := ParseConditionExpression(cfg.Expression); err != nil {
				return err
			}
		}
		if err := ValidateTemplateEnvValues(cfg.TaskPolicy.Env); err != nil {
			return err
		}
	case "for":
		var cfg ForConfig
		if err := decodeConfig(task.Config, &cfg); err != nil {
			return fmt.Errorf("invalid for config: %w", err)
		}
		if strings.TrimSpace(cfg.Var) != "" {
			if err := ValidateEnvName(strings.TrimSpace(cfg.Var)); err != nil {
				return fmt.Errorf("invalid for variable: %w", err)
			}
		}
		if err := ValidateTemplateEnvValues(cfg.TaskPolicy.Env); err != nil {
			return err
		}
	default:
		return fmt.Errorf("unsupported task type %s", task.Type)
	}

	if _, err := ParseTaskPolicy(task.Config); err != nil {
		return err
	}

	return nil
}

func decodeConfig(raw string, target any) error {
	if strings.TrimSpace(raw) == "" {
		raw = "{}"
	}
	return json.Unmarshal([]byte(raw), target)
}

func ValidateTemplateEnvValues(values map[string]string) error {
	for key, value := range values {
		if err := ValidateTemplateString(value); err != nil {
			return fmt.Errorf("invalid template in env %s: %w", key, err)
		}
	}

	return nil
}

File: web/src/app/workflow/validation.ts
import { type Edge } from "@xyflow/react";
import {
  type ConditionConfig,
  type ForConfig,
  type HttpConfig,
  type ShellConfig,
  type TaskConfig,
  type TaskNode,
} from "./types";

export interface WorkflowValidationResult {
  valid: boolean;
  errors: string[];
}

function getTaskLabel(task: TaskNode) {
  return task.name.trim() || task.id;
}

function validatePolicy(config: TaskConfig, task: TaskNode, errors: string[]) {
  const policyFields = [
    ["timeout_seconds", config.timeout_seconds],
    ["retry_count", config.retry_count],
    ["retry_backoff_seconds", config.retry_backoff_seconds],
  ] as const;

  for (const [field, value] of policyFields) {
    if (value === undefined || value === null) {
      continue;
    }
    if (!Number.isFinite(Number(value)) || Number(value) < 0) {
      errors.push(`${getTaskLabel(task)} has an invalid ${field} value.`);
    }
  }
}

function isValidEnvName(value: string) {
  return /^[A-Za-z_][A-Za-z0-9_]*$/.test(value) && !value.startsWith("KINETIC_");
}

export function validateWorkflowDefinition(
  taskNodes: Record<string, TaskNode> | TaskNode[],
  edges: Edge[]
): WorkflowValidationResult {
  const tasks = Array.isArray(taskNodes) ? taskNodes : Object.values(taskNodes);
  const errors: string[] = [];
  const taskMap = new Map<string, TaskNode>();
  const inbound = new Map<string, Edge[]>();
  const outbound = new Map<string, Edge[]>();

  for (const task of tasks) {
    if (!task.id) {
      errors.push("Each task must have an ID.");
      continue;
    }

    taskMap.set(task.id, task);
    inbound.set(task.id, []);
    outbound.set(task.id, []);

    const config = (task.config ?? {}) as TaskConfig;
    validatePolicy(config, task, errors);

    switch (task.type) {
      case "shell":
        if (!((config as ShellConfig).script || "").trim()) {
          errors.push(`${getTaskLabel(task)} requires a shell script.`);
        }
        break;
      case "http":
        if (!((config as HttpConfig).url || "").trim()) {
          errors.push(`${getTaskLabel(task)} requires a request URL.`);
        }
        break;
      case "condition":
        if (!((config as ConditionConfig).expression || "").trim()) {
          errors.push(`${getTaskLabel(task)} requires a condition expression.`);
        }
        break;
      case "for": {
        const forConfig = config as ForConfig;
        if (!Number.isFinite(Number(forConfig.start)) || !Number.isFinite(Number(forConfig.end))) {
          errors.push(`${getTaskLabel(task)} requires numeric start and end values.`);
        }
        const variable = (forConfig.var || "").trim();
        if (variable && !isValidEnvName(variable)) {
          errors.push(`${getTaskLabel(task)} has an invalid loop variable name.`);
        }
        break;
      }
      default:
        errors.push(`${getTaskLabel(task)} has an unsupported task type.`);
        break;
    }
  }

  for (const edge of edges) {
    if (!edge.source || !edge.target) {
      errors.push("Every edge must have both a source and a target.");
      continue;
    }
    if (!taskMap.has(edge.source) || !taskMap.has(edge.target)) {
      errors.push(`Edge ${edge.id || `${edge.source}->${edge.target}`} references a missing task.`);
      continue;
    }

    inbound.get(edge.target)?.push(edge);
    outbound.get(edge.source)?.push(edge);
  }

  for (const task of tasks) {
    if (task.type === "condition") {
      const inEdges = inbound.get(task.id) ?? [];
      const outEdges = outbound.get(task.id) ?? [];
      const handles = new Set(outEdges.map((edge) => edge.sourceHandle).filter(Boolean));

      if (inEdges.length !== 1) {
        errors.push(`${getTaskLabel(task)} must have exactly one inbound edge.`);
      }
      if (outEdges.length !== 2) {
        errors.push(`${getTaskLabel(task)} must have exactly two outbound edges.`);
      }
      if (!handles.has("true") || !handles.has("false") || handles.size !== 2) {
        errors.push(`${getTaskLabel(task)} must connect both true and false branches.`);
      }
      continue;
    }

    if (task.type !== "for") {
      continue;
    }

    const outEdges = outbound.get(task.id) ?? [];
    const bodyEdges = outEdges.filter((edge) => edge.sourceHandle === "body");
    const doneEdges = outEdges.filter((edge) => edge.sourceHandle === "done");
    const unsupportedEdges = outEdges.filter(
      (edge) => edge.sourceHandle !== "body" && edge.sourceHandle !== "done"
    );

    if (bodyEdges.length !== 1) {
      errors.push(`${getTaskLabel(task)} must connect exactly one body branch.`);
    }
    if (doneEdges.length > 1) {
      errors.push(`${getTaskLabel(task)} can connect at most one done branch.`);
    }
    if (unsupportedEdges.length > 0) {
      errors.push(`${getTaskLabel(task)} only supports body and done branches.`);
    }
  }

  const indegree = new Map<string, number>();
  const adjacency = new Map<string, string[]>();

  for (const task of tasks) {
    indegree.set(task.id, 0);
    adjacency.set(task.id, []);
  }

  for (const edge of edges) {
    if (!taskMap.has(edge.source) || !taskMap.has(edge.target)) {
      continue;
    }
    indegree.set(edge.target, (indegree.get(edge.target) ?? 0) + 1);
    adjacency.get(edge.source)?.push(edge.target);
  }

  const queue = tasks
    .map((task) => task.id)
    .filter((id) => (indegree.get(id) ?? 0) === 0);
  let visited = 0;

  while (queue.length > 0) {
    const taskID = queue.shift();
    if (!taskID) {
      continue;
    }
    visited += 1;
    for (const next of adjacency.get(taskID) ?? []) {
      const nextDegree = (indegree.get(next) ?? 0) - 1;
      indegree.set(next, nextDegree);
      if (nextDegree === 0) {
        queue.push(next);
      }
    }
  }

  if (visited !== tasks.length) {
    errors.push("Workflow graph contains a cycle.");
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}
